x/pki/keeper/msg_server_approve_revoke_x_509_root_cert.go

```diff
@@ -51,36 +51,75 @@ func (k msgServer) ApproveRevokeX509RootCert(goCtx context.Context, msg *types.M
 	// check if proposed certificate revocation has enough approvals
 	if len(revocation.Approvals) >= k.CertificateApprovalsCount(ctx, k.dclauthKeeper) {
 		certificates, found := k.GetApprovedCertificates(ctx, msg.Subject, msg.SubjectKeyId)
-		// Assign the approvals to the root certificate
-		for _, cert := range certificates.Certs {
-			if cert.IsRoot {
-				cert.Approvals = revocation.Approvals
-			}
-		}
 		if !found {
 			return nil, pkitypes.NewErrCertificateDoesNotExist(msg.Subject, msg.SubjectKeyId)
 		}
-		k.AddRevokedCertificates(ctx, certificates)
-		k.RemoveApprovedCertificates(ctx, msg.Subject, msg.SubjectKeyId)
-		k.RevokeChildCertificates(ctx, msg.Subject, msg.SubjectKeyId)
-		k.RemoveProposedCertificateRevocation(ctx, msg.Subject, msg.SubjectKeyId)
-
-		// remove from root certs index, add to revoked root certs
 		certID := types.CertificateIdentifier{
 			Subject:      msg.Subject,
 			SubjectKeyId: msg.SubjectKeyId,
 		}
-		k.RemoveApprovedRootCertificate(ctx, certID)
 		k.AddRevokedRootCertificate(ctx, certID)
+		k.RemoveProposedCertificateRevocation(ctx, msg.Subject, msg.SubjectKeyId, msg.SerialNumber)
 
-		// remove from subject -> subject key ID map
-		k.RemoveApprovedCertificateBySubject(ctx, msg.Subject, msg.SubjectKeyId)
-
-		// remove from subject key ID -> certificates map
-		k.RemoveApprovedCertificatesBySubjectKeyID(ctx, msg.Subject, msg.SubjectKeyId)
+		certBySerialNumber, _ := findCertificate(msg.SerialNumber, &certificates.Certs)
+		if certBySerialNumber != nil {
```

Comment:
I think, It would be better to verify that the serial number is not an empty string and If the serial number is non-empty, we can try to find required certificate from the list. If no matching certificate is found, an error should be thrown.